x-pack/filebeat/input/streaming/websocket.go

```diff
@@ -58,6 +65,47 @@ func (l *loggingRoundTripper) RoundTrip(req *http.Request) (*http.Response, erro
 	return resp, err
 }
 
+// heartBeat sends a ping message to the websocket server at regular intervals
+func (k *keepAlive) heartBeat(ctx context.Context, conn *websocket.Conn, start time.Time) context.CancelFunc {
+	ctx, cancel := context.WithCancel(ctx)
+	// set initial read deadline
+	conn.SetReadDeadline(time.Now().Add(k.cfg.KeepAlive.readControlDeadline))
+	// set pong handler to update read deadline
+	conn.SetPongHandler(func(string) error {
+		k.log.Debugw("received pong message from websocket server")
+		k.metrics.pongMessageReceivedTime.Update(time.Since(start).Nanoseconds())
+		return conn.SetReadDeadline(time.Now().Add(k.cfg.KeepAlive.readControlDeadline))
+	})
+
+	// set heartbeat ping routine
+	go func() {
+		ticker := time.NewTicker(k.cfg.KeepAlive.Interval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				if conn != nil {
+					err := conn.WriteControl(websocket.PingMessage, []byte{}, time.Now().Add(k.cfg.KeepAlive.WriteControlDeadline))
+					if err != nil {
+						k.log.Debugw("error sending ping message to websocket server:", err)
+						return
+					}
+					k.log.Debugw("sent ping message to websocket server")
+					k.metrics.pingMessageSendTime.Update(time.Since(start).Nanoseconds())
+				} else {
+					k.log.Debugw("websocket connection is nil but heartbeat routine is still running!")
+				}
+			case <-ctx.Done():
+				k.log.Debugw("heartbeat stopped")
```

Comment:
Move this to the bottom of the function so that it picks up all stops?

Reply:
It's already at the bottom of the select block, do you want the `debug log` to be present at the bottom outside the select ?

Reply:
> do you want the `debug log` to be present at the bottom outside the select ?

Yes. Then is catches all cases of stopping.